Framework/Algorithms/src/AbsorptionCorrectionPaalmanPings.cpp

```diff
@@ -0,0 +1,486 @@
+  m_beamDirection = m_inputWS->getInstrument()->getBeamDirection();
+
+  // Get the input parameters
+  retrieveBaseProperties();
+
+  // Create the output workspace
+  MatrixWorkspace_sptr ass = create<HistoWorkspace>(*m_inputWS);
+  ass->setDistribution(true); // The output of this is a distribution
+  ass->setYUnit("");          // Need to explicitly set YUnit to nothing
+  ass->setYUnitLabel("Attenuation factor");
+  MatrixWorkspace_sptr assc = create<HistoWorkspace>(*m_inputWS);
+  assc->setDistribution(true); // The output of this is a distribution
+  assc->setYUnit("");          // Need to explicitly set YUnit to nothing
+  assc->setYUnitLabel("Attenuation factor");
+  MatrixWorkspace_sptr acc = create<HistoWorkspace>(*m_inputWS);
+  acc->setDistribution(true); // The output of this is a distribution
+  acc->setYUnit("");          // Need to explicitly set YUnit to nothing
+  acc->setYUnitLabel("Attenuation factor");
+  MatrixWorkspace_sptr acsc = create<HistoWorkspace>(*m_inputWS);
+  acsc->setDistribution(true); // The output of this is a distribution
+  acsc->setYUnit("");          // Need to explicitly set YUnit to nothing
+  acsc->setYUnitLabel("Attenuation factor");
+
+  constructSample(m_inputWS->mutableSample());
+
+  const auto numHists = static_cast<int64_t>(m_inputWS->getNumberHistograms());
+  const auto specSize = static_cast<int64_t>(m_inputWS->blocksize());
+
+  // If the number of wavelength points has not been given, use them all
+  if (isEmpty(m_num_lambda))
+    m_num_lambda = specSize;
+  m_xStep = specSize / m_num_lambda; // Bin step between points to calculate
+
+  if (m_xStep == 0) // Number of wavelength points >number of histogram points
+    m_xStep = 1;
+
+  std::ostringstream message;
+  message << "Numerical integration performed every " << m_xStep
+          << " wavelength points";
+  g_log.information(message.str());
+  message.str("");
+
+  // Calculate the cached values of L1, element volumes, and geometry size
+  initialiseCachedDistances();
+  if (m_L1s.empty() || m_containerL1s.empty()) {
+    throw std::runtime_error(
+        "Failed to define any initial scattering gauge volume for geometry");
+  }
+
+  const auto &spectrumInfo = m_inputWS->spectrumInfo();
+  Progress prog(this, 0.0, 1.0, numHists);
+  // Loop over the spectra
+  PARALLEL_FOR_IF(Kernel::threadSafe(*m_inputWS, *ass, *assc, *acc, *acsc))
+  for (int64_t i = 0; i < int64_t(numHists); ++i) {
+    PARALLEL_START_INTERUPT_REGION
+    // Copy over bins
+    ass->setSharedX(i, m_inputWS->sharedX(i));
+    assc->setSharedX(i, m_inputWS->sharedX(i));
+    acc->setSharedX(i, m_inputWS->sharedX(i));
+    acsc->setSharedX(i, m_inputWS->sharedX(i));
+
+    if (!spectrumInfo.hasDetectors(i)) {
+      g_log.information() << "Spectrum " << i
+                          << " does not have a detector defined for it\n";
+      continue;
+    }
+    const auto &det = spectrumInfo.detector(i);
+
+    std::vector<double> sample_L2s(m_numVolumeElements);
```

Comment:
Comments pointing out which terms (e.g. Ass, Acc) these are for would be helpful